behaviour/graql/reasoner/resolution/type-hierarchy.feature

```diff
@@ -0,0 +1,471 @@
+      $v isa subEntity1, has name "a";
+
+      (role1:$x, role2:$z) isa relation2;     # subEntity1 - entity1    -> satisfies rule
+      (role1:$y, role2:$z) isa relation2;     # entity1 - entity1       -> doesn't satisfy rule
+      (role1:$x, role2:$v) isa relation2;     # subEntity1 - subEntity1 -> satisfies rule
+      (role1:$y, role2:$v) isa relation2;     # entity1 - subEntity1    -> doesn't satisfy rule
+      """
+    When materialised keyspace is completed
+    Then for graql query
+      """
+      match
+        $x isa entity1;
+        $y isa entity1;
+        (role1: $x, role2: $y) isa relation1;
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    # Answers are (role1:$x, role2:$z) and (role1:$x, role2:$v)
+    Then answer size in reasoned keyspace is: 2
+    Then for graql query
+      """
+      match
+        $x isa entity1;
+        $y isa entity1;
+        (role1: $x, role2: $y) isa relation1;
+        $y has name 'a';
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 2
+    Then for graql query
+      """
+      match
+        $x isa entity1;
+        $y isa subEntity1;
+        (role1: $x, role2: $y) isa relation1;
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    # Answer is (role1:$x, role2:$v) ONLY
+    Then answer size in reasoned keyspace is: 1
+    Then for graql query
+      """
+      match
+        $x isa entity1;
+        $y isa subEntity1;
+        (role1: $x, role2: $y) isa relation1;
+        $y has name 'a';
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 1
+    Then for graql query
+      """
+      match
+        $x isa subEntity1;
+        $y isa entity1;
+        (role1: $x, role2: $y) isa relation1;
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    # Answers are (role1:$x, role2:$z) and (role1:$x, role2:$v)
+    Then answer size in reasoned keyspace is: 2
+    Then for graql query
+      """
+      match
+        $x isa subEntity1;
+        $y isa entity1;
+        (role1: $x, role2: $y) isa relation1;
+        $y has name 'a';
+      get;
+      """
+    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 2
+    Then materialised and reasoned keyspaces are the same size
+
+
+  Scenario: when a rule is recursive, its inferences respect type hierarchies
+    Given for each session, graql define
+      """
+      define
+
+      entity1 sub entity,
+          has name,
+          plays role1,
+          plays role2;
+
+      subEntity1 sub entity1;
+
+      relation1 sub relation,
+          relates role1,
+          relates role2;
+
+      relation2 sub relation,
+          relates role1,
+          relates role2;
+
+      name sub attribute, value string;
+
+      rule-1 sub rule,
```

Comment:
i do think we can get better names for rules at least, so we don't have to understand the rules fully?
`duplicate-[relation2]-with-specific-players-to-[relation1]` or something

Reply:
Ok I renamed everything.